src/backend/resources.rs

```diff
@@ -110,82 +123,153 @@ fn write_csv(count_map: HashMap<(Id, Binding), u32>, file: &mut OutputFile) {
 // Prints the estimated size (in bits) of the generated hardware along with a breakdown
 // of which primitives contributed to the total number.
 // TODO (priya): Add other primitives
-fn estimated_size(count_map: HashMap<(Id, Binding), u32>) {
+fn estimated_size(count_map: HashMap<(Id, Binding, bool), u32>) {
     let mut estimated_size: u64 = 0;
+    let mut estimated_external_size: u64 = 0;
     eprintln!("Summary of primitives:");
-    for ((name, params), count) in count_map {
+    for ((name, params, is_external), count) in count_map {
         match name.as_ref() {
             "std_reg" => {
-                estimated_size += (count as u64) * (params[0].1);
-                eprintln!(
-                    "{} {} primitive(s) with a bitwidth of {}.",
-                    count, name, params[0].1
-                )
+                if is_external {
+                    estimated_external_size += (count as u64) * (params[0].1);
+                } else {
+                    estimated_size += (count as u64) * (params[0].1);
+                    eprintln!(
+                        "{} {} primitive(s) with a bitwidth of {}.",
+                        count, name, params[0].1
+                    )
+                }
             }
             "std_mem_d1" => {
-                estimated_size += (count as u64) * (params[0].1 * params[1].1);
-                eprintln!(
+                if is_external {
+                    estimated_external_size +=
+                        (count as u64) * (params[0].1 * params[1].1);
+                } else {
+                    estimated_size +=
+                        (count as u64) * (params[0].1 * params[1].1);
```

Comment:
Some code duplication in situations like this could be avoided by putting the RHS into a temporary:

```rs
let size = (count as u64) * (params[0].1 * params[1].1);
if is_external {
  estimated_external_size += size;
} else {
  estimated_size += size;
}
```

Or, it could even make sense to factor out a (nested) function like `add_size(external: bool, amount: u64)` that just contains that `if`.

Reply:
done! I got to learn about Rust closures! would appreciate your thoughts on the refactor :)